src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1916 @@
+        {
+            var previous = _currentToken;
+            _currentToken = _lexer.ScanNextToken(allowTrivia, _options);
+            return previous;
+        }
+
+        /// <summary>
+        /// Given an input text, and set of options, parses out a fully representative syntax tree 
+        /// and list of diagnotics.  Parsing should always succeed, except in the case of the stack 
+        /// overflowing.
+        /// </summary>
+        public static RegexTree TryParse(ImmutableArray<VirtualChar> text, RegexOptions options)
+        {
+            try
+            {
+                // Parse the tree once, to figure out the capture groups.  These are needed
+                // to then parse the tree again, as the captures will affect how we interpret
+                // certain things (i.e. escape references) and what errors will be reported.
+                //
+                // This is necessary as .net regexes allow references to *future* captures.
+                // As such, we don't know when we're seeing a reference if it's to something
+                // that exists or not.
+                var tree1 = new RegexParser(text, options,
+                    ImmutableDictionary<string, TextSpan>.Empty,
+                    ImmutableDictionary<int, TextSpan>.Empty).ParseTree();
+
+                var analyzer = new CaptureInfoAnalyzer(text);
+                var (captureNames, captureNumbers) = analyzer.Analyze(tree1.Root, options);
+
+                var tree2 = new RegexParser(
+                    text, options, captureNames, captureNumbers).ParseTree();
+                return tree2;
+            }
+            catch (Exception e) when (StackGuard.IsInsufficientExecutionStackException(e))
+            {
+                return null;
+            }
+        }
+
+        private RegexTree ParseTree()
+        {
+            var expression = this.ParseAlternatingSequences(consumeCloseParen: true);
+            Debug.Assert(_lexer.Position == _lexer.Text.Length);
+            Debug.Assert(_currentToken.Kind == RegexKind.EndOfFile);
+
+            var root = new RegexCompilationUnit(expression, _currentToken);
+
+            var diagnostics = ArrayBuilder<EmbeddedDiagnostic>.GetInstance();
+            CollectDiagnostics(root, diagnostics);
+
+            return new RegexTree(
+                _lexer.Text, root, diagnostics.ToImmutableAndFree(),
+                _captureNamesToSpan, _captureNumbersToSpan);
+        }
+
+        private static void CollectDiagnostics(RegexNode node, ArrayBuilder<EmbeddedDiagnostic> diagnostics)
+        {
+            foreach (var child in node)
+            {
+                if (child.IsNode)
+                {
+                    CollectDiagnostics(child.Node, diagnostics);
+                }
+                else
+                {
+                    var token = child.Token;
+                    foreach (var trivia in token.LeadingTrivia)
```

Comment:
Should we collect diagnostics from trailing trivia too?

Reply:
regexes only have leading trivia.  i will add comment and assert that that is so here.

Reply:
#resolved